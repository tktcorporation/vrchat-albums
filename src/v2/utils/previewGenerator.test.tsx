import { beforeEach, describe, expect, it, vi } from 'vitest';
import fs from 'fs';
import path from 'path';
import { generatePreviewPng, generatePreviewSvg } from './previewGenerator.tsx';

// XMLSerializerのモック
class MockXMLSerializer {
  serializeToString() {
    return '<svg>test</svg>';
  }
}
global.XMLSerializer = MockXMLSerializer as typeof XMLSerializer;

// DOMのモックを設定
const mockDocument = {
  createElement: vi.fn(),
  createElementNS: vi.fn(),
  fonts: {
    load: vi.fn().mockResolvedValue(undefined), // This mock might be irrelevant now but kept for safety
  },
};
global.document = mockDocument as unknown as Document;

// btoa関数のモック
global.btoa = vi.fn().mockReturnValue('test-base64');

// Image constructorのモック
class MockImage {
  onload = () => {};
  onerror = () => {};
  src = '';
  crossOrigin = '';
  constructor() {
    setTimeout(() => this.onload(), 0); // Simulate async loading
  }
}
global.Image = MockImage as unknown as typeof Image;

// URL.createObjectURLとrevokeObjectURLのモック
global.URL.createObjectURL = vi.fn().mockReturnValue('blob:test');
global.URL.revokeObjectURL = vi.fn();

describe('previewGenerator', () => {
  let fontDataGlobal: ArrayBuffer;

  beforeAll(() => {
    // Load the actual font file once for all tests
    const fontPath = path.resolve(process.cwd(), 'assets/NotoSansCJKjp-Regular.ttf');
    const nodeBuffer = fs.readFileSync(fontPath);
    fontDataGlobal = nodeBuffer.buffer.slice(nodeBuffer.byteOffset, nodeBuffer.byteOffset + nodeBuffer.byteLength);
  });

  beforeEach(() => {
    // モックをリセット
    vi.clearAllMocks();

    // 共通のキャンバスコンテキストのモック (for extractDominantColors and PNG generation)
    const mockCanvasContext = {
      fillStyle: '',
      fillRect: vi.fn(),
      drawImage: vi.fn(),
      getImageData: vi.fn().mockReturnValue({
        data: new Uint8ClampedArray([
          255,0,0,255, 0,255,0,255, 0,0,255,255, // Some basic pixels
          ...Array(1000).fill(0).flatMap(() => [
            Math.floor(Math.random() * 255),
            Math.floor(Math.random() * 255),
            Math.floor(Math.random() * 255),
            255,
          ]),
        ]),
        width: 10, // Simplified dimensions for mock image data
        height: 10,
      }),
    };

    const mockCanvas = {
      width: 800 * 2, // For PNG generation
      height: 600 * 2, // For PNG generation
      getContext: vi.fn().mockImplementation((contextType: string) => {
        if (contextType === '2d') {
          // Mock for extractDominantColors if it needs different dimensions
          if (mockCanvas.width !== 800*2 || mockCanvas.height !== 600*2) {
             return {
                ...mockCanvasContext,
                getImageData: vi.fn().mockReturnValue({
                    data: new Uint8ClampedArray(mockCanvas.width * mockCanvas.height * 4),
                    width: mockCanvas.width,
                    height: mockCanvas.height,
                }),
             };
          }
          return mockCanvasContext;
        }
        return null;
      }),
      toDataURL: vi.fn().mockReturnValue('data:image/png;base64,test123'),
    };

    // Mock for Image used in extractDominantColors
    const mockImageInstance = {
        onload: () => {},
        onerror: () => {},
        src: '',
        crossOrigin: '',
        width: 32, // example width for mock image in extractDominantColors
        height: 32, // example height
    };

    global.Image = vi.fn(() => {
        // Simulate async loading for the image in extractDominantColors
        setTimeout(() => mockImageInstance.onload(), 0);
        return mockImageInstance;
    }) as any;


    mockDocument.createElement.mockImplementation((tagName: string) => {
      if (tagName === 'canvas') {
        // This will be called by extractDominantColors AND generatePreviewPng
        // We need to differentiate or ensure the mock is flexible.
        // For extractDominantColors, canvas width/height are set from image.
        // For generatePreviewPng, they are 800*2, 600*2.
        const newMockCanvas = { ...mockCanvas };
        // Allow width/height to be set on this instance by extractDominantColors
        Object.defineProperty(newMockCanvas, 'width', { writable: true, value: 0 });
        Object.defineProperty(newMockCanvas, 'height', { writable: true, value: 0 });
        return newMockCanvas;
      }
      return {} as HTMLElement;
    });

    // No need to mock createElementNS for this test anymore as Satori generates SVG string directly
  });

  describe('generatePreviewPng', () => {
    it('プレビュー画像を正しく生成できること', async () => {
      const params = {
        worldName: 'Test World PNG',
        imageBase64: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=', // Valid 1x1 PNG
        players: [{ playerName: 'Player 1' }, { playerName: 'Player 2' }],
        showAllPlayers: false,
        fontData: fontDataGlobal,
      };
      const result = await generatePreviewPng(params);
      expect(result).toBe('test123');
      expect(URL.createObjectURL).toHaveBeenCalled();
      expect(URL.revokeObjectURL).toHaveBeenCalledWith('blob:test');
    });

    it('プレイヤーリストが空の場合も正しく処理できること', async () => {
      const params = {
        worldName: 'Test World PNG Empty',
        imageBase64: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=',
        players: null,
        showAllPlayers: false,
        fontData: fontDataGlobal,
      };
      const result = await generatePreviewPng(params);
      expect(result).toBe('test123');
    });

    it('showAllPlayersがtrueの場合、すべてのプレイヤーを表示できること', async () => {
      const params = {
        worldName: 'Test World PNG All Players',
        imageBase64: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=',
        players: Array.from({ length: 20 }, (_, i) => ({ playerName: `Player ${i + 1}` })),
        showAllPlayers: true,
        fontData: fontDataGlobal,
      };
      const result = await generatePreviewPng(params);
      expect(result).toBe('test123');
    });

    it('キャンバスコンテキストの取得に失敗した場合エラーを投げること', async () => {
      mockDocument.createElement.mockImplementation((tagName: string) => {
        if (tagName === 'canvas') {
          return {
            width: 800*2,
            height: 600*2,
            getContext: vi.fn().mockReturnValue(null), // Fail context creation
            toDataURL: vi.fn(),
          };
        }
        return {} as HTMLElement;
      });
      const params = {
        worldName: 'Test World PNG Error',
        imageBase64: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=',
        players: null,
        showAllPlayers: false,
        fontData: fontDataGlobal,
      };
      await expect(generatePreviewPng(params)).rejects.toThrow('Failed to get canvas context');
    });
  });

  describe('generatePreviewSvg', () => {
    // Satori mock is removed. Actual Satori will be used with embedFont: false.
    // fontDataGlobal is loaded in beforeAll and available.

    it('should return SVG string and height, and match snapshot with actual Satori', async () => {
      const params = {
        worldName: 'Test SVG World',
        imageBase64: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=', // Valid 1x1 transparent PNG
        players: [{ playerName: 'SVG Player 1' }],
        showAllPlayers: false,
        fontData: fontDataGlobal, // Use the actual loaded font data
      };
      const result = await generatePreviewSvg(params);
      expect(typeof result.svg).toBe('string');
      // No longer comparing to a fixed mock SVG string.
      expect(result.svg).toMatchSnapshot(); // Snapshot the actual SVG output from Satori
      expect(typeof result.height).toBe('number');
      expect(result.height).toBeGreaterThanOrEqual(600);
    });
  });
});
