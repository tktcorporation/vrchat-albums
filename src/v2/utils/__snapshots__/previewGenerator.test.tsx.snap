// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`previewGenerator > generatePreviewSvg > should return SVG string and height, and match snapshot with actual Satori 1`] = `"<svg width="800" height="600" viewBox="0 0 800 600" xmlns="http://www.w3.org/2000/svg"><mask id="satori_om-id"><rect x="0" y="0" width="800" height="600" fill="#fff"/></mask><rect x="0" y="0" width="800" height="600" fill="#7FB5B5"/><defs><pattern id="satori_biid-0_0" patternContentUnits="userSpaceOnUse" patternUnits="userSpaceOnUse" x="NaN" y="NaN" width="1" height="1"><image x="0" y="0" width="1" height="1" preserveAspectRatio="none" href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII="/></pattern></defs><mask id="satori_om-id-0"><rect x="0" y="0" width="800" height="600" fill="#fff"/></mask><g opacity="0.8"><rect x="0" y="0" width="800" height="600" fill="url(#satori_biid-0_0)"/></g><defs><pattern id="satori_pattern_id-1_0" x="0" y="0" width="1" height="1" patternUnits="objectBoundingBox"><linearGradient id="satori_biid-1_0" x1="0" y1="0" x2="0" y2="1" spreadMethod="pad"><stop offset="0%" stop-color="rgb(59, 130, 246)66"/><stop offset="100%" stop-color="rgb(147, 51, 234)66"/></linearGradient><rect x="0" y="0" width="800" height="600" fill="url(#satori_biid-1_0)"/></pattern></defs><mask id="satori_om-id-1"><rect x="0" y="0" width="800" height="600" fill="#fff"/></mask><rect x="0" y="0" width="800" height="600" fill="url(#satori_pattern_id-1_0)"/><mask id="satori_om-id-2"><rect x="0" y="0" width="800" height="600" fill="#fff"/></mask><mask id="satori_om-id-2-0"><rect x="0" y="0" width="800" height="51" fill="#fff"/></mask><mask id="satori_om-id-2-0-0"><rect x="32" y="24" width="768" height="20" fill="#fff"/></mask><path fill="white" d="M37.1 41.6L38.9 41.6L38.9 28.5L43.4 28.5L43.4 26.9L32.6 26.9L32.6 28.5L37.1 28.5ZM50.2 41.9C51.7 41.9 52.8 41.4 53.8 40.7L53.1 39.5C52.3 40.1 51.5 40.4 50.4 40.4C48.4 40.4 46.9 38.9 46.8 36.6L54.1 36.6C54.2 36.3 54.2 36.0 54.2 35.6C54.2 32.5 52.7 30.5 49.9 30.5C47.4 30.5 45.0 32.7 45.0 36.2C45.0 39.8 47.3 41.9 50.2 41.9ZM46.8 35.3C47.0 33.2 48.4 31.9 49.9 31.9C51.6 31.9 52.6 33.1 52.6 35.3ZM59.7 41.9C62.3 41.9 63.7 40.4 63.7 38.6C63.7 36.6 62.0 35.9 60.4 35.3C59.2 34.9 58.0 34.5 58.0 33.5C58.0 32.6 58.6 31.9 60.0 31.9C60.9 31.9 61.7 32.3 62.5 32.9L63.3 31.7C62.5 31.0 61.3 30.5 60.0 30.5C57.6 30.5 56.2 31.8 56.2 33.5C56.2 35.4 57.9 36.1 59.5 36.7C60.7 37.1 61.9 37.6 61.9 38.7C61.9 39.7 61.2 40.4 59.8 40.4C58.5 40.4 57.5 39.9 56.6 39.2L55.7 40.4C56.7 41.2 58.2 41.9 59.7 41.9ZM69.7 41.9C70.3 41.9 71.1 41.7 71.7 41.5L71.3 40.1C71.0 40.2 70.5 40.4 70.1 40.4C68.8 40.4 68.4 39.6 68.4 38.3L68.4 32.2L71.3 32.2L71.3 30.7L68.4 30.7L68.4 27.7L66.9 27.7L66.7 30.7L65.0 30.8L65.0 32.2L66.6 32.2L66.6 38.2C66.6 40.4 67.4 41.9 69.7 41.9Z M82.5 41.9C85.6 41.9 87.5 40.0 87.5 37.7C87.5 35.5 86.2 34.5 84.5 33.8L82.4 32.9C81.2 32.4 80.0 31.9 80.0 30.4C80.0 29.1 81.0 28.3 82.7 28.3C84.1 28.3 85.1 28.8 86.0 29.7L87.0 28.5C86.0 27.4 84.4 26.7 82.7 26.7C80.0 26.7 78.1 28.3 78.1 30.6C78.1 32.7 79.7 33.7 81.1 34.3L83.2 35.2C84.6 35.9 85.6 36.3 85.6 37.8C85.6 39.3 84.5 40.2 82.5 40.2C81.0 40.2 79.5 39.5 78.5 38.4L77.4 39.7C78.7 41.0 80.4 41.9 82.5 41.9ZM93.1 41.6L95.2 41.6L99.9 26.9L98 26.9L95.6 34.9C95.1 36.6 94.8 38 94.2 39.7L94.1 39.7C93.6 38 93.2 36.6 92.7 34.9L90.3 26.9L88.4 26.9ZM107.6 41.9C109.6 41.9 111.2 41.1 112.2 40.2L112.2 34L107.4 34L107.4 35.5L110.5 35.5L110.5 39.4C109.9 39.9 108.9 40.2 107.8 40.2C104.7 40.2 102.9 37.9 102.9 34.2C102.9 30.6 104.9 28.3 107.8 28.3C109.3 28.3 110.2 28.9 111.0 29.7L112.0 28.5C111.1 27.6 109.8 26.7 107.8 26.7C103.9 26.7 101.0 29.5 101.0 34.3C101.0 39.0 103.8 41.9 107.6 41.9Z M121.8 41.6L124.0 41.6L126.1 32.8C126.4 31.6 126.6 30.5 126.9 29.4L127.0 29.4C127.2 30.5 127.4 31.6 127.7 32.8L129.9 41.6L132.1 41.6L135.2 26.9L133.4 26.9L131.8 34.9C131.6 36.5 131.3 38.1 131 39.7L130.9 39.7C130.6 38.1 130.2 36.5 129.9 34.9L127.8 26.9L126.1 26.9L124.1 34.9C123.7 36.5 123.4 38.1 123.0 39.7L123.0 39.7C122.7 38.1 122.4 36.5 122.1 34.9L120.6 26.9L118.6 26.9ZM141.8 41.9C144.4 41.9 146.8 39.8 146.8 36.2C146.8 32.6 144.4 30.5 141.8 30.5C139.1 30.5 136.7 32.6 136.7 36.2C136.7 39.8 139.1 41.9 141.8 41.9ZM141.8 40.3C139.9 40.3 138.6 38.7 138.6 36.2C138.6 33.7 139.9 32.0 141.8 32.0C143.6 32.0 144.9 33.7 144.9 36.2C144.9 38.7 143.6 40.3 141.8 40.3ZM149.7 41.6L151.5 41.6L151.5 34.6C152.2 32.8 153.3 32.1 154.2 32.1C154.7 32.1 154.9 32.2 155.3 32.3L155.6 30.7C155.3 30.5 154.9 30.5 154.5 30.5C153.2 30.5 152.1 31.3 151.4 32.7L151.3 32.7L151.2 30.7L149.7 30.7ZM159.3 41.9C159.8 41.9 160.1 41.8 160.4 41.7L160.1 40.3C159.9 40.3 159.9 40.3 159.8 40.3C159.5 40.3 159.2 40.1 159.2 39.6L159.2 25.7L157.4 25.7L157.4 39.4C157.4 41.0 158.0 41.9 159.3 41.9ZM166.8 41.9C168.1 41.9 169.3 41.2 170.1 40.3L170.2 40.3L170.3 41.6L171.8 41.6L171.8 25.7L170 25.7L170 29.9L170.1 31.7C169.1 30.9 168.3 30.5 167.0 30.5C164.6 30.5 162.3 32.7 162.3 36.2C162.3 39.8 164.1 41.9 166.8 41.9ZM167.2 40.3C165.3 40.3 164.2 38.8 164.2 36.2C164.2 33.7 165.6 32 167.4 32C168.3 32 169.1 32.3 170 33.1L170 38.8C169.1 39.8 168.2 40.3 167.2 40.3Z "/><mask id="satori_om-id-2-0-1"><rect x="32" y="48" width="200" height="3" fill="#fff"/></mask><path x="32" y="48" width="200" height="3" fill="rgb(79, 70, 229)" d="M33.5,48 h197 a1.5,1.5 0 0 1 1.5,1.5 v0 a1.5,1.5 0 0 1 -1.5,1.5 h-197 a1.5,1.5 0 0 1 -1.5,-1.5 v0 a1.5,1.5 0 0 1 1.5,-1.5"/><clipPath id="satori_cp-id-2-1"><path x="32" y="70" width="736" height="414" d="M44,70 h712 a12,12 0 0 1 12,12 v390 a12,12 0 0 1 -12,12 h-712 a12,12 0 0 1 -12,-12 v-390 a12,12 0 0 1 12,-12"/></clipPath><mask id="satori_om-id-2-1"><rect x="32" y="70" width="736" height="414" fill="#fff"/></mask><image x="32" y="70" width="736" height="414" href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=" preserveAspectRatio="xMidYMid slice" clip-path="url(#satori_cp-id-2-1)" mask="url(#satori_om-id-2-1)"/><mask id="satori_om-id-2-2"><rect x="32" y="504" width="740" height="52" fill="#fff"/></mask><mask id="satori_om-id-2-2-0"><rect x="32" y="504" width="740" height="14" fill="#fff"/></mask><path fill="rgba(255, 255, 255, 0.6)" d="M33.4 516.3L34.7 516.3L34.7 512.2L36.4 512.2C38.6 512.2 40.2 511.2 40.2 509.1C40.2 506.8 38.6 506.1 36.3 506.1L33.4 506.1ZM34.7 511.2L34.7 507.1L36.2 507.1C38.0 507.1 38.9 507.6 38.9 509.1C38.9 510.5 38.0 511.2 36.2 511.2ZM42.9 516.3L48.7 516.3L48.7 515.2L44.3 515.2L44.3 506.1L42.9 506.1ZM52.5 512.1L53.0 510.6C53.4 509.4 53.7 508.3 54.0 507.1L54.1 507.1C54.4 508.3 54.8 509.4 55.1 510.6L55.6 512.1ZM56.9 516.3L58.3 516.3L54.8 506.1L53.4 506.1L49.9 516.3L51.2 516.3L52.2 513.2L55.9 513.2ZM62.1 516.3L63.4 516.3L63.4 512.3L66.5 506.1L65.1 506.1L63.8 509.0C63.5 509.7 63.2 510.4 62.8 511.2L62.7 511.2C62.4 510.4 62.1 509.7 61.7 509.0L60.4 506.1L59.0 506.1L62.1 512.3ZM68.6 516.3L74.6 516.3L74.6 515.2L69.9 515.2L69.9 511.5L73.8 511.5L73.8 510.4L69.9 510.4L69.9 507.2L74.5 507.2L74.5 506.1L68.6 506.1ZM78.8 507.1L80.5 507.1C82.1 507.1 83.0 507.6 83.0 508.9C83.0 510.3 82.1 510.9 80.5 510.9L78.8 510.9ZM84.6 516.3L82.0 511.8C83.4 511.5 84.3 510.5 84.3 508.9C84.3 506.8 82.8 506.1 80.7 506.1L77.5 506.1L77.5 516.3L78.8 516.3L78.8 512.0L80.7 512.0L83.1 516.3ZM89.9 516.5C92.1 516.5 93.4 515.2 93.4 513.6C93.4 512.1 92.5 511.4 91.3 510.8L89.8 510.2C89.1 509.9 88.2 509.5 88.2 508.5C88.2 507.6 88.9 507.0 90.1 507.0C91.0 507.0 91.8 507.4 92.4 508.0L93.1 507.1C92.4 506.4 91.3 505.9 90.1 505.9C88.2 505.9 86.8 507.0 86.8 508.6C86.8 510.1 88.0 510.8 88.9 511.2L90.4 511.9C91.4 512.3 92.1 512.6 92.1 513.7C92.1 514.7 91.3 515.4 90.0 515.4C88.9 515.4 87.9 514.9 87.1 514.1L86.3 515.0C87.2 515.9 88.5 516.5 89.9 516.5Z M101.9 519.1L102.7 518.7C101.5 516.7 100.9 514.3 100.9 512.0C100.9 509.6 101.5 507.2 102.7 505.2L101.9 504.9C100.6 507.0 99.9 509.2 99.9 512.0C99.9 514.7 100.6 517.0 101.9 519.1ZM105.2 516.3L110.9 516.3L110.9 515.3L108.8 515.3L108.8 506.1L107.8 506.1C107.3 506.4 106.6 506.6 105.7 506.8L105.7 507.6L107.5 507.6L107.5 515.3L105.2 515.3ZM113.9 519.1C115.2 517.0 115.9 514.7 115.9 512.0C115.9 509.2 115.2 507.0 113.9 504.9L113.1 505.2C114.3 507.2 114.9 509.6 114.9 512.0C114.9 514.3 114.3 516.7 113.1 518.7Z "/><mask id="satori_om-id-2-2-1"><rect x="32" y="526" width="740" height="30" fill="#fff"/></mask><mask id="satori_om-id-2-2-1-0"><rect x="32" y="526" width="90" height="24" fill="#fff"/></mask><path x="32" y="526" width="90" height="24" fill="rgba(0, 0, 0, 0.3)" d="M44,526 h66 a12,12 0 0 1 12,12 v0 a12,12 0 0 1 -12,12 h-66 a12,12 0 0 1 -12,-12 v0 a12,12 0 0 1 12,-12"/><mask id="satori_om-id-2-2-1-0-0"><rect x="42" y="532" width="70" height="12" fill="#fff"/></mask><path fill="rgba(255, 255, 255, 0.9)" d="M45.6 542.7C47.5 542.7 48.6 541.6 48.6 540.2C48.6 538.9 47.8 538.3 46.8 537.9L45.6 537.3C44.9 537.0 44.1 536.7 44.1 535.9C44.1 535.1 44.8 534.6 45.8 534.6C46.6 534.6 47.2 534.9 47.8 535.4L48.3 534.7C47.7 534.0 46.8 533.6 45.8 533.6C44.2 533.6 43.0 534.6 43.0 535.9C43.0 537.2 44.0 537.8 44.8 538.2L46.0 538.7C46.9 539.1 47.5 539.4 47.5 540.3C47.5 541.2 46.8 541.7 45.7 541.7C44.7 541.7 43.9 541.3 43.2 540.7L42.6 541.4C43.3 542.2 44.4 542.7 45.6 542.7ZM52.0 542.6L53.3 542.6L56.1 533.8L54.9 533.8L53.5 538.5C53.2 539.6 53.0 540.4 52.7 541.4L52.6 541.4C52.3 540.4 52.1 539.6 51.8 538.5L50.3 533.8L49.2 533.8ZM60.7 542.7C61.9 542.7 62.9 542.3 63.4 541.7L63.4 538L60.6 538L60.6 538.9L62.4 538.9L62.4 541.2C62.1 541.6 61.5 541.7 60.8 541.7C58.9 541.7 57.9 540.4 57.9 538.1C57.9 535.9 59.1 534.6 60.8 534.6C61.7 534.6 62.3 534.9 62.7 535.4L63.3 534.7C62.8 534.1 62.0 533.6 60.8 533.6C58.5 533.6 56.8 535.3 56.8 538.2C56.8 541.0 58.4 542.7 60.7 542.7Z M68.2 542.6L69.3 542.6L69.3 539.1L70.8 539.1C72.7 539.1 74.0 538.2 74.0 536.3C74.0 534.4 72.7 533.8 70.7 533.8L68.2 533.8ZM69.3 538.2L69.3 534.7L70.6 534.7C72.1 534.7 72.9 535.1 72.9 536.3C72.9 537.6 72.2 538.2 70.6 538.2ZM76.9 542.7C77.1 542.7 77.3 542.7 77.5 542.6L77.3 541.8C77.2 541.8 77.2 541.8 77.1 541.8C76.9 541.8 76.8 541.7 76.8 541.3L76.8 533.0L75.7 533.0L75.7 541.3C75.7 542.2 76.0 542.7 76.9 542.7ZM80.6 542.7C81.4 542.7 82.2 542.3 82.8 541.8L82.8 541.8L82.9 542.6L83.8 542.6L83.8 538.6C83.8 536.9 83.1 535.9 81.6 535.9C80.5 535.9 79.6 536.3 79.0 536.7L79.4 537.5C79.9 537.1 80.6 536.8 81.4 536.8C82.5 536.8 82.7 537.6 82.7 538.4C80.0 538.7 78.7 539.5 78.7 540.9C78.7 542.0 79.5 542.7 80.6 542.7ZM80.9 541.8C80.3 541.8 79.8 541.5 79.8 540.8C79.8 540.0 80.5 539.4 82.7 539.2L82.7 541.0C82.1 541.5 81.6 541.8 80.9 541.8ZM86.0 545.4C87.3 545.4 88.0 544.4 88.4 543.1L90.9 536.0L89.8 536.0L88.7 539.7C88.5 540.2 88.3 540.9 88.1 541.5L88.0 541.5C87.8 540.9 87.6 540.2 87.4 539.7L86.1 536.0L84.9 536.0L87.6 542.6L87.4 543.1C87.2 543.9 86.7 544.5 85.9 544.5C85.8 544.5 85.6 544.4 85.5 544.4L85.2 545.2C85.4 545.3 85.7 545.4 86.0 545.4ZM94.8 542.7C95.7 542.7 96.4 542.4 96.9 542.0L96.5 541.3C96.1 541.6 95.5 541.8 94.9 541.8C93.7 541.8 92.8 541.0 92.7 539.6L97.1 539.6C97.2 539.4 97.2 539.2 97.2 538.9C97.2 537.1 96.3 535.9 94.6 535.9C93.1 535.9 91.7 537.2 91.7 539.3C91.7 541.5 93.0 542.7 94.8 542.7ZM92.7 538.8C92.9 537.5 93.7 536.8 94.6 536.8C95.6 536.8 96.2 537.5 96.2 538.8ZM98.8 542.6L99.9 542.6L99.9 538.4C100.3 537.3 101.0 536.9 101.5 536.9C101.8 536.9 102.0 536.9 102.2 537.0L102.4 536.0C102.2 535.9 102.0 535.9 101.7 535.9C100.9 535.9 100.3 536.4 99.8 537.2L99.8 537.2L99.7 536.0L98.8 536.0Z M106.1 542.6L110.9 542.6L110.9 541.6L109.2 541.6L109.2 533.8L108.3 533.8C107.8 534.0 107.3 534.2 106.5 534.4L106.5 535.1L108.0 535.1L108.0 541.6L106.1 541.6Z "/></svg>"`;
